File: src/leetcode/linkedlist/GetIntersectionNode.java
package leetcode.linkedlist;

/**
 * @author chengzw
 * @description 相交链表，https://leetcode-cn.com/problems/intersection-of-two-linked-lists/
 * <p>
 * 给你两个单链表的头节点 headA 和 headB ，请你找出并返回两个单链表相交的起始节点。如果两个链表没有交点，返回 null 。
 * 题目数据 保证 整个链式结构中不存在环。
 * 注意，函数返回结果后，链表必须 保持其原始结构。
 * @since 2021/8/28
 */
public class GetIntersectionNode {

    /**
     * 思路：
     * 1.分别计算a链表和b链表的长度
     * 2.让长度长的链表先走差值
     * 3.然后两个指针一起走，遇到相等的节点就是第一个相交的节点
     *
     * @param headA
     * @param headB
     * @return
     */
    public static ListNode solution1(ListNode headA, ListNode headB) {

        //计算链表a的长度
        int lengthA = 0;
        ListNode pa = headA;
        while (pa != null) {
            lengthA++;
            pa = pa.next;
        }

        //计算链表b的长度
        int lengthB = 0;
        ListNode pb = headB;
        while (pb != null) {
            lengthB++;
            pb = pb.next;
        }

        //长的链表的指针先走差值步
        pa = headA;
        pb = headB;
        if (lengthA >= lengthB) {
            for (int i = 0; i < lengthA - lengthB; i++) {
                pa = pa.next;
            }
        } else {
            for (int i = 0; i < lengthB - lengthA; i++) {
                pb = pb.next;
            }
        }

        //两个链表一起走
        while (pa != null && pb != null) {
            if (pa == pb) return pa;
            pa = pa.next;
            pb = pb.next;
        }

        return null;
    }
}

File: go/leetcode/linkedlist/get_intersection_node.go
package linkedlist

/**
 * @description 160. 相交链表
 * @author chengzw
 * @since 2022/5/24
 * @link https://leetcode.cn/problems/intersection-of-two-linked-lists/
 */

/**
给你两个单链表的头节点 headA 和 headB， 请你找出并返回两个单链表相交的起始节点。 如果两个链表不存在相交节点， 返回 null。
题目数据 保证 整个链式结构中不存在环。注意，函数返回结果后，链表必须保持其原始结构 。
*/

/**
思路：
1.分别计算 a 链表和 b 链表的长度
2.让长度长的链表先走差值
3.然后两个指针一起走，遇到相等的节点就是第一个相交的节点

时间复杂度：O(m+n)
空间复杂度：O(1)
*/

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	a := headA
	b := headB
	countA := 0
	countB := 0

	//计算链表 a 的长度
	for a != nil {
		a = a.Next
		countA++
	}

	//计算链表 b 的长度
	for b != nil {
		b = b.Next
		countB++
	}

	//长的链表的指针先走差值步
	a = headA
	b = headB
	if countA > countB {
		count := countA - countB
		for count > 0 {
			a = a.Next
			count--
		}
	} else if countA < countB {
		count := countB - countA
		for count > 0 {
			b = b.Next
			count--
		}
	}
	//两个链表一起走
	for a != nil && b != nil {
		if a == b {
			return a
		}
		a = a.Next
		b = b.Next
	}
	return nil
}
